libraries/platforms/dromajo-vcs/software/src/bp_utils.c

```diff
@@ -0,0 +1,240 @@
+// BlackParrot utilities
+// Compile this file instead of the existing one in the BlackParrot firmware
+
```

Comment:
At a first glance, i thought this should go in bsg_manycore_platform.cpp, because it is how the RISC-V core communicates with the manycore. 

But bsg_manycore_platform is for the simulator environment, in x86? And this is part of the runtime library in RISC-V, right?

Reply:
I guess that's the question. bsg_manycore_platform is for both. When the code is compiled and run, both x86 and RISC-V are able to completely control the manycore. In a real system, the x86 would only be used for bootstrap/debug, but in our systems most likely x86 will be used for program loading and RISC-V will be used for coordination.

But both are "bsg_manycore_platforms"

Reply:
I think I'm more confused, now.

Reply:
bsg_manycore_platform.cpp is purely only for the platform (i.e. RISC-V) to talk to the manycore. @drichmond is right that this file could be combined with bsg_manycore_platform.cpp but I kept them separate in the interest of demarcating what is the overall "platform" code and what is the underlying implementation in terms of writing to specific MMIO addresses etc.

The simulator is code running on x86 natively that handles all the DPI stuff and any host packets.


The final text picture of this is Dromajo runs on x86 but models a RISC-V core that executes the platform. The simulator (also running on x86) aids Dromajo in communicating with the manycore by handling all DPI calls and performs the function of the BlackParrot Host. Based on this picture, what @dpetrisko says should follow about both of them technically being a part of the platform and all that.

Maybe this picture might help
![dromajo_cudalite](https://user-images.githubusercontent.com/17497327/124526267-2ab21680-ddb7-11eb-83ee-f0b240c69934.png)

Reply:
Where is this picture?

Reply:
If I understand correctly, the code in this file should be in platform.cpp. The purpose of platform.cpp is for the implementation of the platform, no? 

I'm still confused

Reply:
It's fine to put in plaform.cpp

Reply:
Relevant functions moved to `bsg_manycore_platform.cpp` for clarity.